test/e2e/httpproxy/backend_tls_test.go

```diff
@@ -0,0 +1,141 @@
+)
+
+func testBackendTLS(namespace string) {
+	Specify("mTLS to backends can be configured", func() {
+		// Backend server cert signed by CA.
+		backendServerCert := &certmanagerv1.Certificate{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "backend-server-cert",
+			},
+			Spec: certmanagerv1.CertificateSpec{
+				Usages: []certmanagerv1.KeyUsage{
+					certmanagerv1.UsageServerAuth,
+				},
+				CommonName: "echo-secure",
+				DNSNames:   []string{"echo-secure"},
+				SecretName: "backend-server-cert",
+				IssuerRef: certmanagermetav1.ObjectReference{
+					Name: "ca-issuer",
+				},
+			},
+		}
+		require.NoError(f.T(), f.Client.Create(context.TODO(), backendServerCert))
+		f.Fixtures.EchoSecure.Deploy(namespace, "echo-secure")
+
+		p := &contourv1.HTTPProxy{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      "backend-tls",
+			},
+			Spec: contourv1.HTTPProxySpec{
+				VirtualHost: &contourv1.VirtualHost{
+					Fqdn: "backend-tls.projectcontour.io",
+				},
+				Routes: []contourv1.Route{
+					{
+						Services: []contourv1.Service{
+							{
+								Name: "echo-secure",
+								Port: 443,
+								UpstreamValidation: &contourv1.UpstreamValidation{
+									CACertificate: "backend-client-cert",
+									SubjectName:   "echo-secure",
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		f.CreateHTTPProxyAndWaitFor(p, httpProxyValid)
+
+		type responseTLSDetails struct {
+			TLS struct {
+				PeerCertificates []string
+			}
+		}
+
+		// Send HTTP request, we will check backend connection was over HTTPS.
+		res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+			Host:      p.Spec.VirtualHost.Fqdn,
+			Condition: e2e.HasStatusCode(200),
+		})
+		require.NotNil(f.T(), res, "request never succeeded")
+		require.Truef(f.T(), ok, "expected 200 response code, got %d", res.StatusCode)
+
+		// Get cert presented to backend app.
+		tlsInfo := new(responseTLSDetails)
+		require.NoError(f.T(), json.Unmarshal(res.Body, tlsInfo))
+		require.Len(f.T(), tlsInfo.TLS.PeerCertificates, 1)
+
+		// Get value of client cert Envoy should have presented.
+		clientSecretKey := client.ObjectKey{Namespace: namespace, Name: "backend-client-cert"}
+		clientSecret := &corev1.Secret{}
+		require.NoError(f.T(), f.Client.Get(context.TODO(), clientSecretKey, clientSecret))
+
+		assert.Equal(f.T(), tlsInfo.TLS.PeerCertificates[0], string(clientSecret.Data["tls.crt"]))
+
+		// Delete client cert so it is rotated.
+		require.NoError(f.T(), f.Client.Delete(context.TODO(), clientSecret))
+		// Make sure it is deleted.
+		require.Eventually(f.T(), func() bool {
+			return errors.IsNotFound(f.Client.Get(context.TODO(), clientSecretKey, clientSecret))
+		}, time.Second*10, time.Millisecond*50)
+		// Make sure it comes back.
+		require.Eventually(f.T(), func() bool {
+			return f.Client.Get(context.TODO(), clientSecretKey, clientSecret) == nil
+		}, time.Second*10, time.Millisecond*50)
```

Comment:
I think there's a race condition here where if it's deleted and recreated in between iterations of the first check for deletion, this will fail. Not sure if it's a real issue in practice or not. One alternative would be to capture the UID before deletion, and then wait until a get succeeds and returns an object with a different UID meaning it was deleted + recreated.

Reply:
ah yeah good point, thats why the old test added an annotation and waited for it to be gone, I'll look into changing this around to prevent races